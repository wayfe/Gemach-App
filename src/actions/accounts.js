import database from "../firebase/firebase";


// add account redux
export const addAccount = (account) => ({
    type: "ADD_ACCOUNT",
    account,
});

// add account firebase
export const startAddAccount = (accountData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            name = "", 
            phoneNum= "", 
            address = "", 
            address2 = "",
            city = "",
            state = "",
            zip = "",
            id = ""
        } = accountData;
        const account = { name, phoneNum, address, address2, city, state, zip, transactions: [] }
        return database.ref(`/users/${uid}/accounts/${id}`).set(account).then(() => {
            dispatch(addAccount({
                id,
                ...account
            }));
        });
    };                                                                                                                       
};

// remove account
export const removeAccount = ({ id } = {}) => ({
    type: "REMOVE_ACCOUNT",
    id
});

// edit account
export const editAccount = (id, updates) => ({
    type: "EDIT_ACCOUNT",
    id,
    updates
});

// edit account firebase
export const startEditAccount = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/accounts/${id}`).update(updates).then(() => {
            dispatch(editAccount(id, updates));
        })
    }
}

// add transactions
export const addTransaction = (id, transaction) => ({
    type: "ADD_TRANSACTION",
    id,
    transaction
});


export const startAddTransaction = (accountId, transaction = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            id = "",
            amount = "",
            note = "",
            date = "",
            type = "",
            payType = "",
            checkNo = "",
            isCleared = false,
        } = transaction
        const transactionData = { amount, note, date, type, payType, isCleared, checkNo}
        return database.ref(`/users/${uid}/accounts/${accountId}/transactions/${id}`).set(transactionData)
        .then(() => {
            dispatch(addTransaction(accountId, transaction));
        });
    };
};

// remove transactions
export const removeTransaction = (accountId, transactionId) => ({
    type: "REMOVE_TRANSACTION",
    accountId,
    transactionId
});

export const startRemoveTransaction = (accountId, transactionId) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/accounts/${accountId}/transactions/${transactionId}`).set(null).then(() => {
            dispatch(removeTransaction(accountId, transactionId));
        });
    }
}

// edit transactions
export const editTransaction = (accountId, transaction) => ({
    type: "EDIT_TRANSACTION",
    accountId,
    transaction
})

export const startEditTransaction = (accountId, transaction) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/accounts/${accountId}/transactions/${transaction.id}`).update(transaction).then(() => {
            dispatch(editTransaction(accountId, transaction));
        })
    }
}

// SET_ACCOUNTS
export const setAccounts = (accounts) => ({
    type: "SET_ACCOUNTS",
    accounts
});


export const startSetAccounts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/accounts/`).once("value").then((snapshots) => {
            const accounts = [];
            snapshots.forEach((i) => {
                const transactionArray = [] 
                const transactions = i.val().transactions;

                for (var key in transactions) {
                    transactionArray.push({...transactions[key], id: key});
                }
                    
                const account = {...i.val(), id: i.key, transactions: transactionArray}
                    
                accounts.push(account);
            });
            console.log(accounts);
            dispatch(setAccounts(accounts));
        });
    };                                                                                                                       
};
