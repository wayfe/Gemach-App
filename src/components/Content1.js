import React from "react";

const ContentOne = () => (
    <div>
        <h1>this is content 1</h1>
    </div>
);

export default ContentOne;