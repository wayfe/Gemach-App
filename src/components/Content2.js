import React from "react";

const ContentTwo = () => (
    <div>
        <h1>this is content 2</h1>
    </div>
);

export default ContentTwo;