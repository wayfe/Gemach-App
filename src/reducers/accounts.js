
const AccountsReducerDefaultState = [];

export default (state = AccountsReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_ACCOUNT":
            return [
                ...state, action.account
            ];
        case "REMOVE_ACCOUNT":
            return state.filter(({ id }) => id !== action.id);
        case "EDIT_ACCOUNT":
            return state.map((account) => {
                if (account.id === action.id) {
                    return {
                        ...account, ...action.updates
                    };
                } else {
                    return account
                }
            });
        case "ADD_TRANSACTION":
        return state.map((account) => {
            if (account.id === action.id) {
                return {
                    ...account, transactions: [...account.transactions, action.transaction]
                };
            } else {
                return account
            }
        });
        case "REMOVE_TRANSACTION":
        return state.map((account) => {
            if (account.id === action.accountId) {
                return {
                    ...account, transactions: account.transactions.filter((transaction) => transaction.id !== action.transactionId)
                };
            } else {
                return account
            }
        });
        case "EDIT_TRANSACTION":
        return state.map((account) => {
            if (account.id === action.accountId) {
                return {
                    ...account, transactions: [...account.transactions.filter((transaction) => transaction.id !== action.transaction.id), action.transaction]
                };
            } else {
                return account
            }
        });
        case "SET_ACCOUNTS":
            return action.accounts;
        default:
            return state;
    }
};
