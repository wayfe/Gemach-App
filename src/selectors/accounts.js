
//Get Visible Data
export default (accounts, {text}) => {
    return accounts.filter((account) => {
        const textMatch = account.name.toLowerCase().includes(text.toLowerCase());

        return textMatch;
    }).sort((a,b) => {return a.name > b.name ? 1 : -1})
}

