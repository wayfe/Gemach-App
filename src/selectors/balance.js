export const getBal = (transactions) => {

    const deposits = transactions.filter(transaction => transaction.type === "deposit");
    const withdrawals = transactions.filter(transaction => transaction.type === "withdrawal");

    const incoming =  deposits === [] ? 0 : deposits
        .map((transaction) => transaction.amount)
        .reduce((accumulator, currentVal) => accumulator + currentVal, 0);

    const outgoing = withdrawals === [] ? 0 : withdrawals
        .map((transaction) => transaction.amount)
        .reduce((accumulator, currentVal) => accumulator + currentVal, 0);

    return incoming - outgoing;
};

export const getPendingBal = (transactions) => {
    const pendingTransactions = transactions.filter(transaction => transaction.isCleared === false);

    const deposits = pendingTransactions.filter(transaction => transaction.type === "deposit");
    const withdrawals = pendingTransactions.filter(transaction => transaction.type === "withdrawal");

    const incoming =  deposits === [] ? 0 : deposits
        .map((transaction) => transaction.amount)
        .reduce((accumulator, currentVal) => accumulator + currentVal, 0);

    const outgoing = withdrawals === [] ? 0 : withdrawals
        .map((transaction) => transaction.amount)
        .reduce((accumulator, currentVal) => accumulator + currentVal, 0);

    return incoming - outgoing;

    return incoming - outgoing;
};