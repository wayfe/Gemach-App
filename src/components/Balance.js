import React from "react";
import numeral from 'numeral';
import {connect} from "react-redux";
import {getBal, getPendingBal} from "../selectors/balance";


const Balance = (props) => (
    <div>
        <p>Available Balance: {numeral((props.totalBalance - props.pendingBalance) /100).format('$0,0.00')}</p>
        <p>Current Balance: {numeral(props.totalBalance /100).format('$0,0.00')}</p>
        <p>Total Pending Transactions: {numeral(props.pendingBalance /100).format('$0,0.00')}</p>
    </div>
)

const mapStateToProps = (state, props) => {
    const transactions = props.transactions ? props.transactions : [].concat.apply([], state.accounts.map((account) => account.transactions))
    return {
        totalBalance: getBal(transactions),
        pendingBalance: getPendingBal(transactions)
    }
}

export default connect(mapStateToProps)(Balance);
