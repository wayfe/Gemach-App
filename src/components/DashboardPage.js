import React from "react";
import DataListFilters from "./Filters/DataListFilters";
import AccountList from "./Accounts/AccountList";
import Balance from "./Balance";

const DashboardPage = () => (
    <div>
        <Balance />
        <DataListFilters 
            activeFilters={["text"]}
        />
        <AccountList />
    </div>
);

export default DashboardPage;