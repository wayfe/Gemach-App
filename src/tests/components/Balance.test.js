import React from "react";
import {shallow} from "enzyme"
import {Balance} from "../../components/Balance";
import {getBal, getPendingBal} from "../../selectors/balance";
import accounts from "../fixtures/accounts";

test("should render balance with transaction balances", () => {
    const transactions = accounts[0].transactions;
    const wrapper = shallow(<Balance totalBalance={getBal(transactions)} pendingBalance={getPendingBal(transactions)}/>);
    expect(wrapper).toMatchSnapshot();
});
