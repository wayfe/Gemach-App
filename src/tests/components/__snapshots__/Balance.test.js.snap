// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render balance with transaction balances 1`] = `
<div>
  <p>
    Available Balance: 
    $80.00
  </p>
  <p>
    Current Balance: 
    $220.00
  </p>
  <p>
    Total Pending Transactions: 
    $140.00
  </p>
</div>
`;
